Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp

```diff
@@ -240,6 +244,31 @@ void DeferredWorkTimer::cancelPendingWorkSafe(JSGlobalObject* globalObject)
         setTimeUntilFire(0_s);
 }
 
+void DeferredWorkTimer::cancelPendingWork(VM& vm)
+{
+    ASSERT(vm.heap.isInPhase(CollectorPhase::End));
+    Locker locker { m_taskLock };
+
+    auto isValid = [&](auto& ticket) {
+        for (JSCell* dependency : ticket->dependencies()) {
+            if (!vm.heap.isMarked(dependency))
+                return false;
+        }
+        return vm.heap.isMarked(ticket->target()->globalObject()) && vm.heap.isMarked(ticket->scriptExecutionOwner());
+    };
+
+    bool needToFire = false;
+    for (auto& ticket : m_pendingTickets) {
+        if (!ticket->isCancelled() && !isValid(ticket)) {
+            // At this point, no one can visit the dependencies. So, it's safe to clear.
+            ticket->cancelAndClear();
+            needToFire = true;
+        }
+    }
+    if (needToFire && !isScheduled() && !m_currentlyRunningTask)
+        setTimeUntilFire(0_s);
```

Comment:
The ticket may have its corresponding task pending in the `m_tasks`. Since `DeferredWorkTimer::doWork` handles all cases I think we should just let it remove all related components. I think this makes code clear and consistent.

Reply:
If it's in `m_tasks` then it should already be scheduled, I think.

Reply:
Yes, that's why we check `!isScheduled()` here.